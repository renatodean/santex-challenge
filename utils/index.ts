import { Item } from "../types/product";

/**
 * From a given product it will return the price as a formated string
 * @param product
 * @param quantity
 * @returns string
 */
export const getPrice = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return ((productPrice * (quantity || 1)) / 100).toFixed(2);
};

/**
 * From a given product it will return the price as number
 * @param product
 * @param quantity
 * @returns number
 */
export const getPriceNumber = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return (productPrice * (quantity || 1)) / 100;
};

/**
 * From a given query from router it will return the query value
 * @param query
 * @returns string
 */
export const getQueryValue = (query: any) => {
  if (Array.isArray(query)) {
    return query[0];
  } else {
    return query || "";
  }
};
