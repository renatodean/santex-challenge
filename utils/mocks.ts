export const mockedProducts = {
  products: {
    items: [
      {
        __typename: "Product",
        id: "1",
        name: "Laptop",
        slug: "laptop",
        description:
          "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
        featuredAsset: {
          __typename: "Asset",
          id: "1",
          name: "derick-david-409858-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 129900 },
          { __typename: "ProductVariant", price: 139900 },
          { __typename: "ProductVariant", price: 219900 },
          { __typename: "ProductVariant", price: 229900 },
        ],
      },
      {
        __typename: "Product",
        id: "2",
        name: "Tablet",
        slug: "tablet",
        description:
          'If the computer were invented today, what would it look like? It would be powerful enough for any task. So mobile you could take it everywhere. And so intuitive you could use it any way you wanted — with touch, a keyboard, or even a pencil. In other words, it wouldn’t really be a "computer." It would be Tablet.',
        featuredAsset: {
          __typename: "Asset",
          id: "2",
          name: "kelly-sikkema-685291-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/5a/kelly-sikkema-685291-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/b8/kelly-sikkema-685291-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 32900 },
          { __typename: "ProductVariant", price: 44500 },
        ],
      },
    ],
  },
};

export const mockedEmptyProducts = { products: { items: [] } };

export const mockedProduct = {
  product: {
    __typename: "Product",
    id: "1",
    name: "Laptop",
    slug: "laptop",
    description:
      "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
    featuredAsset: {
      __typename: "Asset",
      id: "1",
      name: "derick-david-409858-unsplash.jpg",
      type: "IMAGE",
      source:
        "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
      preview:
        "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
    },
    variants: [
      { __typename: "ProductVariant", price: 129900 },
      { __typename: "ProductVariant", price: 139900 },
      { __typename: "ProductVariant", price: 219900 },
      { __typename: "ProductVariant", price: 229900 },
    ],
  },
};

export const mockedStorageState = {
  showSnackbar: true,
  products: [
    {
      product: mockedProduct.product,
      quantity: 3,
    },
  ],
};
