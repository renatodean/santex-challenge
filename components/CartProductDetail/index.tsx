import React from "react";
import styled from "@emotion/styled";
import { Item } from "../../types/product";
import { getPrice } from "../../utils";

const Container = styled.li``;

const ProductRow = styled.div`
  display: grid;
  grid-template-columns: 2fr 1fr;
  margin-bottom: 6px;
`;

const ProductInfo = styled.div`
  font-size: 16px;
  font-weight: 400;
`;

const ProductPrice = styled.div`
  font-size: 16px;
  font-weight: 600;
  text-align: right;
`;

interface CartProductDetailProps {
  product: Item;
  quantity: number;
}

const CartProductDetail = ({ product, quantity }: CartProductDetailProps) => (
  <Container>
    <ProductRow>
      <ProductInfo>
        {quantity > 1 ? `${quantity}x ` : ""}
        {product.name}
      </ProductInfo>
      <ProductPrice>$ {getPrice(product, quantity)}</ProductPrice>
    </ProductRow>
  </Container>
);

export default CartProductDetail;
