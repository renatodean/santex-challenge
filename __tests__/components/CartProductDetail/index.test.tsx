import { render } from "@testing-library/react";
import CartProductDetail from "../../../components/CartProductDetail";
import { mockedProduct } from "../../../utils/mocks";

describe("CartProductDetail snapshot", () => {
  it("renders the CartProductDetail", () => {
    const { container } = render(
      <CartProductDetail product={mockedProduct.product} quantity={3} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the CartProductDetail with quantity 1", () => {
    const { container } = render(
      <CartProductDetail product={mockedProduct.product} quantity={1} />
    );
    expect(container).toMatchSnapshot();
  });
});
