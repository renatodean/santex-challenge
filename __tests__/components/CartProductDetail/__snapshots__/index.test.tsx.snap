// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartProductDetail snapshot renders the CartProductDetail 1`] = `
<div>
  <li
    class="css-0"
  >
    <div
      class="css-27h8oo"
    >
      <div
        class="css-19dhafh"
      >
        3x 
        Laptop
      </div>
      <div
        class="css-6gq9lb"
      >
        $ 
        3897.00
      </div>
    </div>
  </li>
</div>
`;

exports[`CartProductDetail snapshot renders the CartProductDetail with quantity 1 1`] = `
<div>
  <li
    class="css-0"
  >
    <div
      class="css-27h8oo"
    >
      <div
        class="css-19dhafh"
      >
        Laptop
      </div>
      <div
        class="css-6gq9lb"
      >
        $ 
        1299.00
      </div>
    </div>
  </li>
</div>
`;
